### PR TITLE
Connects the carrier account settings form to the API

### LINKED ISSUE #2052: Connect carrier account settings form with the new carriers endpoint
The form implemented in https://github.com/Automattic/woocommerce-services/pull/2049 should be using the new Carriers endpoint https://github.com/Automattic/woocommerce-connect-server/issues/1362

### PR DESCRIPTION
**Pull request by Ferdev**
Closes #2052

Sends all the data in the carrier account's settings form to the WooCommerce Connect server to register the new carrier in EasyPost.

In order to test it, get the account number and postal code from the "Woo UPS testing account" in the secret store. For the rest of the address, use the Automattic company address from the FG (but omit the postal code and use the one in the secret store).

### CODE REVIEW

**Comment by harriswong** [comment]
👀

File: classes/class-wc-connect-api-client.php [hunk 1]
```diff
@@ -283,6 +283,16 @@ public function send_shipping_label_refund_request( $label_id ) {
 			return $this->request( 'POST', '/shipping/label/' . $label_id . '/refund' );
 		}
 
+		/**
+		 * Request a refund for a given shipping label
```
Comment (c-shultz) [inline classes/class-wc-connect-api-client.php:287]: Header text needs an update

**Comment by Ferdev** [comment]
> It like maybe there's a new file that didn't get committed: `classes/class-wc-rest-connect-shipping-carriers-controller.php`
Fixed! Sorry, I don't know why that file wasn't added to the commit!

File: client/extensions/woocommerce/woocommerce-services/lib/initialize-carrier-accounts-state.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+		'city',
+		'state',
+		'country',
+		'postal_code',
+		'phone',
+		'email',
+		'name',
+		'title',
+		'website',
+		'license_agreement',
```
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/lib/initialize-carrier-accounts-state.js:21]: Do we still need `license_agreement`? I don't see this used in the API post payload spec.

File: client/extensions/woocommerce/woocommerce-services/lib/initialize-carrier-accounts-state.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	];
```
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/lib/initialize-carrier-accounts-state.js:22]: I think we need `company` in the payload.

File: client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/reducer.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+reducers[ WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_DISCONNECT_CARRIER ] = ( state, { carrier } ) => {
+	const settings = state[ carrier ].settings;
+
+	const newState = {
+		...state,
+		[ carrier ]: { settings },
+	};
+
+	return newState;
+};
```
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/reducer.js:144]: Is the `newState` here the same as the input `state`? For disconnect, do we need to call an API?

File: client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/test/selectors.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+					postal_code: true,
+					state: true,
+				},
+				requiredFields: [
+					'account_number',
+					'name',
+					'address',
+					'city',
+					'state',
+					'state',
```
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/test/selectors.js:59]: Duplicated.

File: client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/test/selectors.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+					license_agreement: true,
+					name: true,
+					phone: true,
+					postal_code: true,
+					state: true,
+				},
+				requiredFields: [
+					'account_number',
+					'name',
+					'address',
```
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/test/selectors.js:56]: This should be `street1`

File: client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/test/selectors.js [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+		const result = getCarrierAccountsState( state, siteId, carrier );
+		expect( result ).to.eql( {
+			settings: {
+				fieldErrors: {},
+				ignoreValidation: {
+					account_number: true,
+					address: true,
+					city: true,
+					company_name: true,
+					company_website: true,
```
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/test/selectors.js:43]: I think `company_name` and `company_website` no longer match the fields.

File: client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/test/selectors.js [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+					'postal_code',
+					'phone',
+					'email',
+					'company_name',
+					'job_title',
+					'company_website',
+					'license_agreement',
+				],
+				showUPSInvoiceFields: false,
+				values: { country: 'US' },
```
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/test/selectors.js:70]: Maybe good to include `isSaving` as well.

File: client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/list-item.js [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+	const connectCarrierHandler = () => {
+		const url = new URL( window.location.href );
+		url.searchParams.set( 'carrier', data.carrierId );
+		window.location.href = url.href;
+	}
+
+	const renderActions = ( credentials ) => {
+		const connectButton = () => {
+			return <Button compact onClick={ connectCarrierHandler }>
```
Comment (harriswong) [inline client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/list-item.js:72]: If we have button style with `<a>`, we can then avoid the onclick handler

**Comment by harriswong** [comment]
In a smaller window, the "Connect your UPS account" session seems to be squeezed to the left:
image

**Comment by harriswong** [comment]
When I filled out the form except the UPS invoice 90 days checkbox, I get the following error:
image

It may not be related to this PR but have you seen this before?

File: client/extensions/woocommerce/app/settings/shipping/index.js [hunk 10]
```diff
@@ -41,6 +41,7 @@ class Shipping extends Component {
 	};
 
 	render = () => {
+		alert( 'entra' )
```
Comment (c-shultz) [inline client/extensions/woocommerce/app/settings/shipping/index.js:44]: 😄

**Comment by c-shultz** [comment]
I also see the same `"settings" is not allowed` message

**Comment by c-shultz** [comment]
Ah, I see. WCS is adding a `settings` field to every request: https://github.com/Automattic/woocommerce-services/blob/develop/classes/class-wc-connect-api-client.php#L414 

The Connect Server will need to accept that in the request validators.

**Comment by c-shultz** [comment]
I updated the endpoint validation on this PR: https://github.com/Automattic/woocommerce-connect-server/pull/1413

File: client/extensions/woocommerce/woocommerce-services/lib/initialize-carrier-accounts-state.js (hunk 2 above)
Comment (Ferdev) [reply client/extensions/woocommerce/woocommerce-services/lib/initialize-carrier-accounts-state.js:21]: It was needed in the form (it was a required field) but it isn't in the API, so I was removing it in the action that sends the form: https://github.com/Automattic/woocommerce-services/blob/5d87e9e2e22d6feddde09b02698e3dd30bac0d77/client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/actions.js#L71

But we've removed the field in the form, so I'll get rid of all the references of this field.

File: client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/reducer.js (hunk 4 above)
Comment (Ferdev) [reply client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/reducer.js:144]: Yes, we have a PR that will take care of it https://github.com/Automattic/woocommerce-services/pull/2067

File: client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/list-item.js (hunk 9 above)
Comment (Ferdev) [reply client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/list-item.js:72]: Yeah, better that way indeed.

**Comment by Ferdev** [comment]
> In a smaller window, the "Connect your UPS account" session seems to be squeezed to the left:

This one has been fixed in the previous PR: #2049

**Comment by Ferdev** [comment]
> I updated the endpoint validation on this PR: [Automattic/woocommerce-connect-server#1413](https://github.com/Automattic/woocommerce-connect-server/pull/1413)

Oh yes, I completely forgot that I had to include that settings property to the validators in WooCommerce Connect Server. Thanks for taking care of it in that PR!! :heart:

File: client/extensions/woocommerce/app/settings/shipping/index.js (hunk 10 above)
Comment (Ferdev) [reply client/extensions/woocommerce/app/settings/shipping/index.js:44]: :innocent: :rofl:

Comment (Ferdev) [reply client/extensions/woocommerce/app/settings/shipping/index.js:44]: Also: fixed in #2049

**Comment by Ferdev** [comment]
I've merged develop into this branch and fixed a bug with the website check. I think this PR is ready for another review!

**Comment by c-shultz** [comment]
This is working well for me and looks good overall! The only issue I see is that some JS unit tests need to be cleaned up. I think we need to prioritize merging this PR, so we can get synced up. I'll open an issue for the unit tests.

**Pull request merged**